OpenSSL/crypto.py

```diff
@@ -25,13 +25,43 @@
 TYPE_DSA = _lib.EVP_PKEY_DSA
 
 
+
 class Error(Exception):
     """
     An error occurred in an `OpenSSL.crypto` API.
     """
 
 
+
+def _exception_from_context_error(exception_type, store_ctx):
+    """
+    Convert a :py:func:`OpenSSL.crypto.verify_cert` failure into a Python
+    exception.
+
+    When a call to native OpenSSL X509_verify_cert fails, additonal information
+    about the failure can be contained from the store context.
+    """
+
+    errors = [
+        _lib.X509_STORE_CTX_get_error(store_ctx._store_ctx),
+        _lib.X509_STORE_CTX_get_error_depth(store_ctx._store_ctx),
+        _native(_ffi.string(_lib.X509_verify_cert_error_string(_lib.X509_STORE_CTX_get_error(store_ctx._store_ctx)))),
+    ]
+    _x509 = _lib.X509_STORE_CTX_get_current_cert(store_ctx._store_ctx)
+    if _x509 != _ffi.NULL:
```

Comment:
Oh, jeez, good catch.

Admittedly, I'm not sure how I can trigger this case without some pretty contrived mocking. I'm not inclined to remove the check, though, because the documents at https://www.openssl.org/docs/crypto/X509_verify_cert.html are pretty unclear what "exceptional circumstances" might be.

What if we just do something like

     _x509 = _lib.X509_STORE_CTX_get_current_cert(self._store_ctx)                                                                                                                                             
    if _x509 != _ffi.NULL:                                                                                                                                                                                    
        _cert = _lib.X509_dup(_x509)                                                                                                                                                                            
        pycert = X509.__new__(X509)                                                                                                                                                                             
        pycert._x509 = _ffi.gc(_cert, _lib.X509_free)                                                                                                                                                           
        return X509StoreContextError(errors, pycert) 
    else:
        Error('Exception context error?')